set.mm

```diff
@@ -7594,6 +7594,14 @@ sufficient to develop all propositional calculus of the logic (together with
       ( wo pm1.4 syl ) BAEABECBAFDG $.
   $}
 
+  ${
+    orcomdd.1 $e |- ( ph -> ( ps -> ( ch \/ th ) ) ) $.
+    $( Commutativity of logic disjunction, in double deduction form.
+       (Contributed by Giovanni Mascellani, 19-Mar-2018.) $)
+    orcomdd $p |- ( ph -> ( ps -> ( th \/ ch ) ) ) $=
+      ( wo pm1.4 syl6 ) ABCDFDCFECDGH $.
+  $}
```

Comment:
This is simply orcomd (up to writing "ph /\ ps") and not much used, so I would not add it.

Reply:
This is a general issue: I just moved this theorem from a mathbox to main, so it is not added (does not increase the size of set.mm). There is a rule that "If you need a theorem from another mathbox, typically it is moved to the main part of set.mm.".  I obeyed this rule for a long time, and thought at least Norm did support/encourage this. This is, in my opinion, why the option /i (or /INCLUDE_MATHBOXES) is available for the minimize command. Since my mathbox is nearly at the end of set.mm, I often detect cases where my proofs can be shortened by using theorems of other's mathboxes, using this option. Then I move these to main according to the above mentioned rule. I do not want to spend a lot of time to investigate if these theorems pay off in the usual way how new theorems are rated. Only in actually obscure cases I revert such minimizations.

If this proceding is not acceptable (anymore), I will not use option /I in the future anymore, resulting in no or at least less theorems being moved from mathboxes to main. @david-a-wheeler @glacode @giomasce what do you think?

Reply:
I agree, and that's my approach. When mmj2 grabs (by automation or search) a theorem from another mathbox, I first prepare a PR for the moveToMainPart (of all the needed theorems), and then a second PR for mathbox only stuff.

I guess that yamma (or other tools), at some point, could have a $allowmathboxes directive and/or a diagnostic warning/error if the diagnostic is missing, and you're using a theorem from another mathbox.
But I don't see this being a priority.

Reply:
There are 4 (additional) proofs which could be shortened using ~orcomdd: ~3orel2, ~elicoelioo, ~nneo, ~swoer.
Is this a sufficient argument to move/keep ~orcomdd in main?

Reply:
I do not know yet. Maybe @digama0 can chime in.  There is a convention that when proving in "deduction style", one uncurries the antecedents, that is, one writes one antecedent as a big conjunction instead of keeping nested implications. That way, a proof part in deduction style typically uses things like https://us.metamath.org/mpeuni/imp.html at the beginning, then uses a bunch of (single) deduction style lemmas, and then if necessary uses https://us.metamath.org/mpeuni/ex.html to go back. Of course, if one adds a whole library of double deduction lemmas, then this will shorten the proof, but at the cost of doubling the size of the library.  And this can be done in a "slippery slope" way: one adds a first double deduction lemma, which will make more proofs shortenable by adding a second double deduction lemma, etc.

Therefore, double deductions should be added only in the most elementary parts of set.mm (typically, formulas with only `->`, `-.`, `<->`).  It's not clear where to draw the line, and it's true that commutation of disjunction is pretty basic too. But the sections with conjunction and disjunction are already very large, so I'm bit skeptical.

Reply:
I think it is fine to justify these kind of theorems on a strict "does it decrease character count" basis. That is enough to block any slippery slopes.

Reply:
@digama0 : what about the "proof style" I described above ?  Also, I am not sure that the "does it decrease character count" is enough to block slippery slopes, since the addition of one such theorem may make new theorems fulfill the criterion, etc.

As for proof style, this is already illustrated in the current proof of orcomdd:
```
 9   syl6.1=orcomdd.1 $e |- ( ph -> ( ps -> ( ch \/ th ) ) )
12   syl6.2=pm1.4     $p |- ( ( ch \/ th ) -> ( th \/ ch ) )
13 orcomdd=syl6     $p |- ( ph -> ( ps -> ( th \/ ch ) ) )
```
which is indeed shorter than the more systematic proof
```
16       imp.1=orcomdd.1 $e |- ( ph -> ( ps -> ( ch \/ th ) ) )
17     orcomd.1=imp    $p |- ( ( ph /\ ps ) -> ( ch \/ th ) )
18   ex.1=orcomd     $p |- ( ( ph /\ ps ) -> ( th \/ ch ) )
19 orcomdd=ex      $p |- ( ph -> ( ps -> ( th \/ ch ) ) )
```
For easy proofs, shorter is better since anyway both are readable, but for longer proofs, maybe a more systematic method is preferable over a highly optimized but less uniform hence harder to grasp proof. This is analogous to computer programs.

Reply:
I too like the style where all the conditions are an antecedent joined by conjunctions. Not sure how hard a line I'd want to take on this, however. This is far from our first double deduction theorem,[1] which doesn't automatically say it is fine to add more, but does say we are partway down the slope whether it is slippery or otherwise.

Anyone have any appetite for putting a discouraged tag on `orcomdd` and having the comment suggest `orcomd` (plus `imp` and `ex` where needed) instead? Although part of me says that sort of thing should be decided by the existing rules around what makes proofs shorter, rather than by trying to form opinions about whether double deduction is good or bad.

[1] https://us.metamath.org/mpeuni/syl6.html instead of https://us.metamath.org/mpeuni/syl.html , https://us.metamath.org/mpeuni/syl6mpi.html instead of https://us.metamath.org/mpeuni/mpisyl.html , probably a fair number of others I'm not finding just now.

Reply:
> [1] https://us.metamath.org/mpeuni/syl6.html instead of https://us.metamath.org/mpeuni/syl.html , https://us.metamath.org/mpeuni/syl6mpi.html instead of https://us.metamath.org/mpeuni/mpisyl.html , probably a fair number of others I'm not finding just now.

Note that these are in implicational calculus, and fall into the category I mentioned above (that is: at most `->`, `<->`, `-.`).

Reply:
I think the proposal of @jkingdon is a good compromise for now: I will putting a discouraged tag on orcomdd (name it orcomddOLD) and having the comment suggest orcomd. Fuethermore, I will replace all occurences of orcomdd in proofs (only three) by orcomd. If nobody objects/intervenes within one year, this theorem would be deleted.

In the future, I will be more careful moving theorems from mathboxes to main...

Reply:
I have problems with ~sbcom3OLD, which uses ~orcomdd: It is commented out (by Norm) with the following comment:

> Note: this proof requires many
>        hours to "MM> PROVE sbcom3OLD" and "MM-PA> MINIMIZE_WITH *" and
>        "MM-PA> SAVE NEW_PROOF".  The discouragements will prevent accidental
>        minimization attempts and other operations that would interfere with
>        scripts.  For the time being, we are keeping this proof as a test case,
>        but it will be deleted eventually. -NM $)

After activating this theorem, I was not able to open it neither with metamath.exe nor with mmj2. Therefore, I cannot replace ~orcomdd. Do we really need this theorem? If yes, we have to keep ~orcomdd (in GM's mathbox).

Reply:
> After activating this theorem, I was not able to open it neither with metamath.exe nor with mmj2. Therefore, I cannot replace ~orcomdd. Do we really need this theorem?

I don't know, I'm not sure what it is a test case for. Like is the idea that maybe a future tool could open the proof more quickly than "many hours"?

Reply:
That infamous theorems has been causing problems already. Since it is superseded by sbcom3 in every aspect, I would remove it.
If one needs "stress tests" for verifiers, there are other places for that.

Reply:
Sorry, I have little time this week. Feel free to resolve this one.